sdk/attestation/azure-security-attestation/src/private/crypto/openssl/opensslcert.cpp

```diff
@@ -0,0 +1,451 @@
+    {
+      openssl_asn1_time notAfterTime(
+          make_openssl_unique(ASN1_TIME_adj, nullptr, expirationTime, 0, 0));
+
+      if (X509_set1_notAfter(certificate.get(), notAfterTime.get()) != 1)
+      {
+        throw OpenSSLException("X509_set1_notAfter"); // LCOV_EXCL_LINE
+      }
+    }
+
+    { // Add the subject Key ID - this is the thumbprint of the public key. Note that we have to
+      // have called X509_set_pubkey before this call.
+      auto extension = CreateExtensionFromConfiguration(
+          certificate, certificate, NID_subject_key_identifier, "hash");
+      if (X509_add_ext(certificate.get(), extension.get(), -1) != 1)
+      {
+        throw OpenSSLException("X509_add_ext"); // LCOV_EXCL_LINE
+      }
+    } // namespace _internal
+
+    // Add the authority Key ID - Note that this needs to be done *after* setting the subject key
+    // identifier.
+    {
+      auto extension(make_openssl_unique(X509_EXTENSION_new));
+      if (issuer)
+      {
+        extension = CreateExtensionFromConfiguration(
+            certificate, issuer, NID_authority_key_identifier, "keyid:always");
+      }
+      else
+      {
+        extension = CreateExtensionFromConfiguration(
+            certificate, certificate, NID_authority_key_identifier, "keyid:always");
+      }
+      if (X509_add_ext(certificate.get(), extension.get(), -1) != 1)
+      {
+        throw OpenSSLException("X509_add_ext"); // LCOV_EXCL_LINE
+      }
+    }
+
+    {
+      auto extension = CreateExtensionFromConfiguration(
+          certificate,
+          certificate,
+          NID_key_usage,
+          (issuer ? "critical,keyCertSign,digitalSignature" : "critical,keyCertSign"));
+      if (X509_add_ext(certificate.get(), extension.get(), -1) != 1)
+      {
+        throw OpenSSLException("X509_add_ext");
+      }
+    }
+
+    // Export the key to sign the certificate.
+    {
+      std::unique_ptr<Cryptography::AsymmetricKey> privateKey;
+      {
+        auto exportedPrivateKey = signingKey->ExportPrivateKey();
+        privateKey = Cryptography::ImportPrivateKey(exportedPrivateKey);
+      }
+      // We know that Crypto::ImportPublicKey always returns an OpenSSLAsymmetricKey.
+      // This is a bit of a hack but it's an acceptable assumption to make.
+      OpenSSLAsymmetricKey* key = static_cast<OpenSSLAsymmetricKey*>(privateKey.get());
+      if (X509_sign(certificate.get(), key->GetKey().get(), EVP_sha256()) == 1)
+      {
+        throw OpenSSLException("X509_sign"); // LCOV_EXCL_LINE
+      }
+    }
+
+    return certificate;
+  } // namespace _internal
+
+  std::unique_ptr<Cryptography::X509Certificate> OpenSSLX509Certificate::CreateFromPrivateKey(
+      std::unique_ptr<Cryptography::AsymmetricKey> const& key,
+      std::string const& subjectName)
+  {
+    auto utcTime(time(nullptr));
+    time_t expirationTime;
+    {
+      // This certificate expires in 8 hours.
+      struct std::tm currentTime;
+#if _WINDOWS
```

Comment:
nit.

our general guideline is to use `#if defined(MACRO)` instead of `#if MACRO`.   

Also, how do we know `_WINDOWS` is always there on Windows?  
We use `#if defined(_WIN32)` in our platform header from Azure-core (sdk/core/azure-core/inc/azure/core/platform.hpp),  you might want to follow that.

You might want to include platform.hpp and use `#if defined(AZ_PLATFORM_WINDOWS)`